fix: edge case about sync

SWvheerden opened the pull request:
Description
---
This fixes an edge case around sync where the local node has a higher chain header, but the remote node has a higher actual block chain height. 

Motivation and Context
---
When doing sync, and determining from where to sync, we need to look at the where out actual block chain is, especially when comparing chains. 
On the current code its possible that you node has a valid header chain of higher proof of work then the remote syncing peer, but the blocks are not valid for that chain, thus the remote chain has an actual higher proof of work block chain. 
The current code will ban the peer for lying about the pow, as the header sync uses the last chain header to determine pow, while the listing state will use the chain_meta_data to determine pow. 

We should look at out chain header tip when determining the fork header before starting header sync, but when we compare the chains, we need to look at what our actual pow chain is and not the tip chain header. 


How Has This Been Tested?
---
Manual. 

What process can a PR reviewer use to test or verify this change?
---

<!-- Checklist -->
<!-- 1. Is the title of your PR in the form that would make nice release notes? The title, excluding the conventional commit
tag, will be included exactly as is in the CHANGELOG, so please think about it carefully. -->


Breaking Changes
---

- [x] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [ ] Other - Please specify

<!-- Does this include a breaking change? If so, include this line as a footer -->
<!-- BREAKING CHANGE: Description what the user should do, e.g. delete a database, resync the chain -->

SWvheerden committed:
fix edge case about sync

hansieodendaal reviewed (commented):
The revised code, although it may be correct, is really confusing. For example, there seems to be no difference in meaning between the words `local_tip_header` and `local_blockchain_tip_header`. When looking at the code, the significance of the change is also lost. For example, why would these be different?

`let local_tip_header = self.db.fetch_last_chain_header().await?;`
vs.
`let local_blockchain_tip_header = self.db.fetch_chain_header(self.local_metadata.height_of_longest_chain()).await?;`

`let header_tip_height = local_tip_header.height();`
vs.
`let header_tip_height = local_blockchain_tip_header.height();`

SWvheerden commented:
I am all ears if you have better names for them. 
but when syncing we have two important terms here: 
tip_header: the highest height header we have.
blockchain_tip_header: the height header we have a full block for, aka forms a blockchain. 

And the reason this is important is how the sync process works. 
1. We download headers only and add them to the main chain. 
2. We download the blocks for those headers and add them to the main chain. 

And it very much means that when entering header sync we have more headers than the current pee, but they have a larger block height than us. Which means we need to rewind our chain and sync to thiers.

SWvheerden committed:
fmt

stringhandler reviewed (changes requested):
I don't agree with this change. It feels incomplete. 

It is very confusing to pass both the highest header and the highest block into a method and then use one for some rules and one for the others.

I think it could lead to future bugs. 

Even if this does solve the problem, I would like a test to make sure that this is the case, so that this edge case is not forgotten about

SWvheerden commented:
The comments here point out the exact problem here, the sync code is stupidly complex. But it is all the edge cases vs. performance. 

We don't use one highest header for one rule, and the highest block for another,  we use both, always. 
The highest header we use to determine the split height. 
The highest block we use to determine pow. 


The much simpler approach would be to clear all the changes made after a failed header sync, but from previous discussions, the majority of the team is against that for performance reasons. And its valid, if you download most of the chain, and fail on connection, then you have to restart from scratch. So it is a valid objection, but for that to work, and be secure, we need complex code to solve this. As it stands, this opens up the node to a valid attack which will cause the node to be stuck forever, and ban the whole network.

SWvheerden committed:
change names

SWvheerden committed:
rename header sync status

SWvheerden committed:
snall fix

Commit:
Add consistency to `attempt_sync`

Added consistency to `attempt_sync` i.t.o. naming, variable re-use and use of information.

SWvheerden committed:
Merge pull request #36 from hansieodendaal/ho_attempt_sync

feat: add consistency to attempt_sync

SWvheerden commented:
closing this for: https://github.com/tari-project/tari/pull/5837
this code is contained in that PR + extra fixes
